solr/core/src/java/org/apache/solr/filestore/PackageStoreAPI.java

```diff
@@ -0,0 +1,273 @@
+      }
+      CryptoKeys cryptoKeys = null;
+      try {
+        cryptoKeys = new CryptoKeys(keys);
+      } catch (Exception e) {
+        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+            "Error parsing public keyts in ZooKeeper");
+      }
+      for (String sig : sigs) {
+       if(cryptoKeys.verify(sig, buf) == null){
+         throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Signature does not match any public key : " +sig );
+       }
+
+      }
+    }
+
+  }
+
+  @EndPoint(
+      path = "/node/files/*",
+      method = SolrRequest.METHOD.GET,
+      permission = PermissionNameProvider.Name.FILESTORE_READ_PERM)
+  public class FSRead {
+    @Command
+    public void read(SolrQueryRequest req, SolrQueryResponse rsp) {
+      String path = req.getPathTemplateValues().get("*");
+      String pathCopy = path;
+      String getFrom = req.getParams().get("getFrom");
+      if (getFrom != null) {
+        coreContainer.getUpdateShardHandler().getUpdateExecutor().submit(() -> {
+          log.debug("Downloading file {}", pathCopy);
+          try {
+            packageStore.fetch(pathCopy, getFrom);
+          } catch (Exception e) {
+            log.error("Failed to download file: " + pathCopy, e);
+          }
+          log.info("downloaded file : {}", pathCopy);
+        });
+        return;
+
+      }
+      if (path == null) {
+        path = "";
+      }
+
+      PackageStore.FileType typ = packageStore.getType(path);
+      if (typ == PackageStore.FileType.NOFILE) {
+        rsp.add("files", Collections.singletonMap(path, null));
+        return;
+      }
+      if (typ == PackageStore.FileType.DIRECTORY) {
+        rsp.add("files", Collections.singletonMap(path, packageStore.list(path, null)));
+        return;
+      }
+      if (req.getParams().getBool("meta", false)) {
+        if (typ == PackageStore.FileType.FILE) {
+          int idx = path.lastIndexOf('/');
+          String fileName = path.substring(idx + 1);
+          String parentPath = path.substring(0, path.lastIndexOf('/'));
+          List l = packageStore.list(parentPath, s -> s.equals(fileName));
+          rsp.add("files", Collections.singletonMap(path, l.isEmpty() ? null : l.get(0)));
+          return;
+        }
+      } else {
+        writeRawFile(req, rsp, path);
+      }
+    }
+
+    private void writeRawFile(SolrQueryRequest req, SolrQueryResponse rsp, String path) {
+      ModifiableSolrParams solrParams = new ModifiableSolrParams();
+      solrParams.add(CommonParams.WT, FILE_STREAM);
+      req.setParams(SolrParams.wrapDefaults(solrParams, req.getParams()));
+      rsp.add(FILE_STREAM, (SolrCore.RawWriter) os -> {
+        packageStore.get(path, (it) -> {
+          try {
+            org.apache.commons.io.IOUtils.copy(it.getInputStream(), os);
+          } catch (IOException e) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "error reading file" + path);
```

Comment:
Please change:
    "error reading file"
to:
    "Error reading file: "
(Notice capitalization and additional space at end of string).